From dd706cc52de81531978039ff8cbd2162aecc4f29 Mon Sep 17 00:00:00 2001
From: Amber Lin <Amber.Lin@amd.com>
Date: Fri, 22 Jul 2016 14:56:10 -0400
Subject: [PATCH 1826/2094] drm/amdgpu: Get user pages in non-current task

get_user_pages is changed to be called from current task where mm is being
operated on. get_user_pages_remote is introduced when the target mm is not
the current task's. Adjust amdgpu_ttm_tt_get_user_pages so it'll get user
pages in both cases.

Change-Id: If0e306c107c3663e2eb0d900009021265e97e04d
Signed-off-by: Amber Lin <Amber.Lin@amd.com>

 Conflicts:
	drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c
Signed-off-by: Avinash M N <avimn@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c
index 12fe6ba..60c3ef6 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_ttm.c
@@ -675,7 +675,12 @@ int amdgpu_ttm_tt_get_user_pages(struct ttm_tt *ttm, struct page **pages)
 		list_add(&guptask.list, &gtt->guptasks);
 		spin_unlock(&gtt->guptasklock);
 
-		r = get_user_pages(userptr, num_pages, flags, p, NULL);
+		if (gtt->usertask->mm == current->mm)
+			r = get_user_pages(userptr, num_pages, flags, p, NULL);
+		else
+			r = get_user_pages_remote(gtt->usertask,
+					gtt->usertask->mm, userptr, num_pages,
+					flags, p, NULL);
 
 		spin_lock(&gtt->guptasklock);
 		list_del(&guptask.list);
-- 
2.7.4

